src/ILCompiler.Compiler/src/Compiler/Compilation.cs

```diff
@@ -184,9 +184,20 @@ public void AddCompilationRoot(MethodDesc method, string reason, string exportNa
             public void AddCompilationRoot(TypeDesc type, string reason)
             {
                 if (type.IsGenericDefinition)
+                {
                     _graph.AddRoot(_factory.NecessaryTypeSymbol(type), reason);
+                }
                 else
+                {
                     _graph.AddRoot(_factory.ConstructedTypeSymbol(type), reason);
+
+                    // If the type has a thread static field then we should eagerly create a helper
+                    // to access such fields at runtime. This is required for multi-module compilation.
+                    if (type.IsDefType && (((DefType)type).ThreadStaticFieldSize > 0))
```

Comment:
@nattress Is this a more general problem?

E.g. I have a class Foo with a static field X in module A that I never touch in module A. Then I have a module B that touches it. We compile both in multifile mode. A.obj won't contain the static base symbol because it wasn't used. B.obj will contain a reference that doesn't resolve.